binaries/runtime/src/operator/shared_lib.rs

```diff
@@ -203,6 +203,10 @@ impl<'lib> SharedLibraryOperator<'lib> {
                     input: None,
                     stop: false,
                 },
+                IncomingEvent::Reload => {
+                    // TODO: Reload shared library.
+                    continue;
+                }
```

Comment:
> [quite difficult for Rust code](https://fasterthanli.me/articles/so-you-want-to-live-reload-rust) 

That's an excellent article. I naively thought that we could just swap context just like in Python.

So, instead of reloading shared library, what about re-spawning a Custom node or Runtime, copying the previous state somewhere and passing it to the newly spawned Custom node or Runtime.

Reply:
The issue is that the state is controlled entirely by the operator. The operator gives us a opaque pointer on initialization and the runtime just passes this pointer on `on_event` calls. Only the operator itself knows what to do with it.

So we cannot just copy the memory referenced by the pointer since we don't know the size of the data. It might even contain pointers to other data internally that is part of the state too. Another challenge is that the new operator might use a different data format, so copying the old state in could result in invalid values and undefined behavior.

The only way to copy the state over is with cooperation between the operator, the runtime, and the new operator. We could for example introduce a `serialize_state` event that needs to be handled manually by the operator code. The operator needs to serialize all state into some common format (e.g. JSON) and the new operator needs to be able to deserialize and apply this state again. It also needs to be able to migrate the data from old structures in case the data format changed. Since this approach is quite complex and requires manual work, I think it's only suited for some planned updates (e.g. update a running operator from version 1 to version 1.1), but not for hot reloading...

Reply:
Agreed, thanks for the detailed explanation! Let's try to see if people use this feature for python, and how they use it, and what state management is more beneficial for the user.